application/slack_messenger/slack_messenger.py

```diff
@@ -0,0 +1,176 @@
+    env = filter_env[0]["value"] if len(filter_env) > 0 else ""
+    new_state = message["NewStateValue"]
+    alarm_description = message["AlarmDescription"]
+    trigger = message["Trigger"]
+    project_id = f"{namespace}-{env}"
+    log_groups = [
+        f"{project_id}-event-processor",
+        f"{project_id}-event-sender",
+        f"{project_id}-cr-fifo-dlq-handler",
+        f"{project_id}-fifo-dlq-handler",
+    ]
+    filters = {"report_key": get_report_key(metric_name)}
+    color = "warning"
+
+    if message["NewStateValue"] == "ALARM":
+        color = "#e01e5a"
+    elif message["NewStateValue"] == "OK":
+        color = "good"
+    link = (
+        "https://console.aws.amazon.com/cloudwatch/home"
+        f"?region={region}#alarm:alarmFilter=ANY;name={quote(alarm_name.encode('utf-8'))}"
+    )
+    slack_message = {
+        "blocks": [
+            {"type": "section", "text": {"type": "mrkdwn", "text": f":rotating_light:  *<{link}|{alarm_name}>*"}}
+        ],
+        "attachments": [
+            {
+                "color": color,
+                "fields": [
+                    {"title": "Alarm Name", "value": alarm_name, "short": True},
+                    {"title": "Alarm State", "value": new_state, "short": True},
+                    {"title": "Alarm Description", "value": alarm_description, "short": False},
+                    {
+                        "title": "Trigger",
+                        "value": f"{trigger['Statistic']} {metric_name} {trigger['ComparisonOperator']} "
+                        f"{str(trigger['Threshold'])} for {str(trigger['EvaluationPeriods'])} period(s) "
+                        f" of {str(trigger['Period'])} seconds.",
+                        "short": False,
+                    },
+                    {
+                        "title": "",
+                        "value": f"<{generate_cloudwatch_url(region, log_groups, filters)}|View Logs>",
+                    },
+                ],
+                "ts": timestamp,
+            }
+        ],
+    }
+    return slack_message
+
+
+def get_report_key(metric_name):
+    if metric_name == "InvalidPostcode":
+        return INVALID_POSTCODE_REPORT_ID
+    elif metric_name == "InvalidOpenTimes":
+        return INVALID_OPEN_TIMES_REPORT_ID
+    return ""
+
+
+def send_msg_slack(message):
+    url = environ["SLACK_WEBHOOK_URL"]
+    channel = environ["SLACK_ALERT_CHANNEL"]
+    headers: Dict[str, str] = {"Content-Type": "application/json", "Accept": "application/json"}
+
+    message["channel"] = channel
+    message["icon_emoji"] = ""
+
+    resp = post(
+        url=url,
+        headers=headers,
+        json=message,
+    )
+    logger.info(
+        "Message sent to slack",
+        extra={"slack_message": message, "status_code": resp.status_code, "response": resp.text},
+    )
+
+
+def generate_cloudwatch_url(region: str, log_groups: list, filters: dict, limit: int = 100):
+    def escape(s):
```

Comment:
can we not use urllib.parse.quote instead of rolling our own here?

Reply:
I tried with the urllib but it didn't work. AWS  URL expects a different format

Reply:
check this solution out using urllib, it appears that aws use double encoding with some substitution, will avoid you having these funky s1 s2 3 etc constants

Reply:
Sure